scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkCounters.scala

```diff
@@ -0,0 +1,89 @@
+package com.twitter.scalding.spark_backend
+
+import com.twitter.scalding.{ExecutionCounters, StatKey}
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.util.AccumulatorV2
+
+import scala.collection.mutable.{Map => MutableMap}
+
+/**
+ * We use a new SparkCounters instance for each SparkWriter to keep track of scalding counters
+ */
+class SparkCounters(sparkSession: SparkSession, counterPrefix: String = java.util.UUID.randomUUID.toString) {
+
+  val accumulator: SparkCountersInternal = new SparkCountersInternal
+
+  /**
+   * Register for spark to be aware of accumulator. Note that accumulators are now automatically garbage
+   * collected: https://github.com/apache/spark/pull/4021 so we don't need to explicitly release them.
+   */
+  {
+    sparkSession.sparkContext.register(accumulator, counterPrefix)
+  }
+
+  /**
+   * We return an ExecutionCounters interface. The results will not be stable unless spark execution has
+   * finished.
+   */
+  def asExecutionCounters(): ExecutionCounters = {
+    val copyState = MutableMap[StatKey, Long]()
+    for (kvPair <- accumulator.value) {
+      val keyEntry = StatKey(kvPair._1._2, kvPair._1._1)
+      copyState.getOrElseUpdate(keyEntry, 0L)
+      copyState(keyEntry) += kvPair._2
+    }
+    new ExecutionCounters {
+      def keys = copyState.keys.toSeq.toSet
+      def get(key: StatKey) = copyState.get(key)
+    }
+  }
+
+}
+
+sealed class SparkCountersInternal(
+    mapState: MutableMap[(String, String), Long] = MutableMap[(String, String), Long]()
+) extends AccumulatorV2[((String, String), Long), MutableMap[(String, String), Long]]
+    with Serializable {
+
+  /**
+   * Should be used inside of spark executor actions for a side-effect free counter implementation.
+   */
+  def add(in: Iterator[((String, String), Long)]): Unit =
+    while (in.hasNext) {
+      val count = in.next
+      mapState.getOrElseUpdate(count._1, 0L)
+      mapState(count._1) += count._2
+    }
+
+  // internal for spark do not call
+  override def add(in: ((String, String), Long)): Unit = {
+    mapState.getOrElseUpdate(in._1, 0L)
+    mapState(in._1) += in._2
+  }
+
+  // internal for spark do not call
+  override def value() = mapState
+
+  // internal for spark do not call
+  override def copy(): SparkCountersInternal = {
+    val copyState = MutableMap[(String, String), Long]()
+    for (kvPair <- mapState) {
```

Comment:
this isn't thread-safe: mapState can change while iterating.

Can you comment why this is safe?

Alternatives:
1. use `mapState.synchronized { }` before reading/writing
2. use an `immutable.Map[(String, String), Long]` and an `AtomicReference` so taking a copy is very cheap (although updates will be maybe twice as slow, so maybe the mutex approach above is better).

Reply:
all of these methods on accumulator operate on a "share nothing" model. The driver maintains the "true" state of the accumulator and merges updates from other accumulator states returned from executors. I am 99% sure that the merging process is thread safe. 

If you look at the implementation for what is passed to executors: 
```scala
  // Called by Java when serializing an object
  final protected def writeReplace(): Any = {
    if (atDriverSide) {
      if (!isRegistered) {
        throw new UnsupportedOperationException(
          "Accumulator must be registered before send to executor")
      }
      val copyAcc = copyAndReset()
      assert(copyAcc.isZero, "copyAndReset must return a zero value copy")
      val isInternalAcc = name.isDefined && name.get.startsWith(InternalAccumulator.METRICS_PREFIX)
      if (isInternalAcc) {
        // Do not serialize the name of internal accumulator and send it to executor.
        copyAcc.metadata = metadata.copy(name = None)
      } else {
        // For non-internal accumulators, we still need to send the name because users may need to
        // access the accumulator name at executor side, or they may keep the accumulators sent from
        // executors and access the name when the registered accumulator is already garbage
        // collected(e.g. SQLMetrics).
        copyAcc.metadata = metadata
      }
      copyAcc
    } else {
      this
    }
  }
``` 

executor threads receive a state-less version of the accumulator.

Reply:
the only time we read the value of the driver accumulator is at the end after execution has already ended and all updates to the accumulator have been merged by spark.

Reply:
Okay, can you add that comment to the code around the class something like:

```scala
/**
 * regarding thread-safety: .... 
 * permalink to this code: ... 
```

Reply:
hmmm maybe not I'm starting to doubt based on the official example: https://github.com/apache/spark/blob/0494dc90af48ce7da0625485a4dc6917a244d580/core/src/main/scala/org/apache/spark/util/AccumulatorV2.scala#L451

I want to try to write a unit test to prove if this is necessary because synchronizing threads is going to be a big perf hit.

Reply:
Sync won't be a big perf hit: you are copying here. The added cost of synchronization will not be large. Also this probably happens very rarely (at some kind of boundary, such as the task finishing). I would be surprised if you could measure a hit.

Reply:
good point this only happens after an rdd actions starts/returns. So assuming a very complicated job has ~100-1000 rdd actions this is no problem

 ok let's just do it better safe than sorry